helix-term/src/commands/lsp.rs

```diff
@@ -1324,3 +1331,191 @@ pub fn select_references_to_symbol_under_cursor(cx: &mut Context) {
+                    // in the document
+                    doc.reset_text_annotations(view_id);
+                    return;
+                }
+            };
+
+            let mut hints = match response {
+                Some(h) if !h.is_empty() => h,
+                _ => {
+                    doc.reset_text_annotations(view_id);
+                    view.last_inlay_hint_request_id = None;
+                    return;
+                }
+            };
+
+            let offset_encoding = match doc.language_server() {
+                Some(ls) => ls.offset_encoding(),
+                None => return,
+            };
+
+            // Most language servers will already send them sorted but ensure this is the case to
+            // avoid errors on our end.
+            hints.sort_unstable_by_key(|inlay_hint| inlay_hint.position);
+
+            let mut annotations = TextAnnotations::default();
+
+            let default_style = editor.theme.find_scope_index("ui.virtual.inlay-hint");
+            let param_style = editor
+                .theme
+                .find_scope_index("ui.virtual.inlay-hint.parameter");
+            let type_style = editor.theme.find_scope_index("ui.virtual.inlay-hint.type");
+
+            let doc_text = &doc.text();
+
+            for hint in hints {
+                let char_idx =
+                    match helix_lsp::util::lsp_pos_to_pos(doc_text, hint.position, offset_encoding)
+                    {
+                        Some(pos) => pos,
+                        // Skip inlay hints that have no "real" position
+                        None => continue,
+                    };
+
+                let label = match hint.label {
+                    lsp::InlayHintLabel::String(s) => s,
+                    lsp::InlayHintLabel::LabelParts(parts) => parts
+                        .into_iter()
+                        .map(|p| p.value)
+                        .collect::<Vec<_>>()
+                        .join(""),
+                };
+
+                let style = match hint.kind {
+                    Some(lsp::InlayHintKind::TYPE) => type_style,
+                    Some(lsp::InlayHintKind::PARAMETER) => param_style,
+                    // We can't warn on unknown kind here since LSPs are free to set it or not, for
+                    // example Rust Analyzer does not: every kind will be `None`.
+                    _ => default_style,
+                };
+
+                if let Some(true) = hint.padding_left {
+                    annotations.add_inline_annotations(
+                        Rc::new([InlineAnnotation::new(char_idx, " ")]),
+                        None,
+                    );
+                }
+
+                annotations.add_inline_annotations(
+                    Rc::new([InlineAnnotation::new(char_idx, label)]),
+                    style.map(helix_core::syntax::Highlight),
+                );
```

Comment:
layers carray a certain amount of overhead. Instead of creating a seperate layer for the each annotation create a `Vec<InlineAnnotation>` and then convert that to `Rc<[InlineAnnotation]>` (same for everwhere else you create a layer, there should be only one layer per type of virtual text, one for padding, one for each kind of label (paramters,  types and default style))

Reply:
Done